__1.For-Loops with range fuctions and with Operators__

- In Python, a for loop is used to iterate over a sequence (that can be a list, tuple, string, or any other iterable object) and execute a block of code for each item in the sequence.
- The loop starts by taking the first item in the sequence and assigning it to the variable item.

- The block of code inside the loop is then executed with the current value of item.

- After the block of code is executed, the loop moves on to the next item in the sequence and assigns it to item.

- The loop continues this process until it has gone through all the items in the sequence.

In [1]:
for i in range(10):    #the for loop iterates sereies of numbers using range fuction.
    print(i)

0
1
2
3
4
5
6
7
8
9


In [4]:
for i in range(0, 11, 2):    #Printing even numbers from 0 to 10.
    print(i)

0
2
4
6
8
10


In [3]:
fruits = ["apple", "banana", "cherry"]
for fruit in fruits:                  # the for loop iterates over a list of fruits and prints each fruit on a new line.
    print(fruit)

apple
banana
cherry


In [1]:
names = ["Manohar", "Arun kumar", "Rahul sekar"]
for name in names:              # the for loop iterates over a list of names and prints each name on a new line.
    print(name)

Manohar
Arun kumar
Rahul sekar


In [6]:
#1.Sum of a list
numbers = [1, 2, 3, 4, 5]
sum = 0

for num in numbers:
    sum += num

print(f"The sum is: {sum}")

The sum is: 15


In [7]:
#2.String Concatenation
words = ["Hello", "world", "!"]
sentence = ""

for word in words:
    sentence += word + " "

print(f"The sentence is: {sentence}")

The sentence is: Hello world ! 


In [2]:
# for loop with range() function and addition operator
for i in range(5):
    print(i + 1)

1
2
3
4
5


In [3]:
# for loop with range() function and multiplication operator
for i in range(1, 6):
    print(i * 2)

2
4
6
8
10


In [1]:
#Tables using range fuction
num = int(input("Enter the number you want to table: "))
for i in range(1, 11):
 print(f"{num} x {i} = {num * i}")

Enter the number you want to table: 5
5 x 1 = 5
5 x 2 = 10
5 x 3 = 15
5 x 4 = 20
5 x 5 = 25
5 x 6 = 30
5 x 7 = 35
5 x 8 = 40
5 x 9 = 45
5 x 10 = 50


In [4]:
#Using zip function, we can add the index position numbers.
list1 = [1, 2, 3, 4, 5]
list2 = [1, 2, 3, 4, 5]
list3 = [2, 2, 3, 4, 5]
result = []
for x, y, z in zip(list1, list2, list3):
    result.append(x + y + z)
print(result)

[4, 6, 9, 12, 15]


__2.Loops with Enumeration__

- In addition to iterating over the elements of a sequence, you can also use a for-loop with enumeration to keep track of the index of each element. The enumerate() function in Python can be used to iterate over a sequence and keep track of the index of each element. 

- the enumerate function is used to iterate over the fruits list, and for each element, it returns a tuple containing the index of the element and the element itself. The for statement then iterates over each tuple, and the variables i and fruit are assigned to the index and the element, respectively, for each iteration of the loop. The code block within the loop then prints the value of i and fruit for each iteration.


In [9]:
fruits = ['Manohar', 'arun', 'mallesh']
for i, fruit in enumerate(fruits):
    print(i, fruit)

0 Manohar
1 arun
2 mallesh


In [11]:
fruits = ['rahul', 'pavithra', 'manish', 'ramya', 'rakshitha', 'sneha']
for i, fruit in enumerate(fruits, 1):    #Second argument specifies that the index should start from 1 instead of 0
    print(i, fruit)

1 rahul
2 pavithra
3 manish
4 ramya
5 rakshitha
6 sneha


__3.Nested for loops__

- Nested for-loops are when one for-loop is inside another for-loop, allowing you to iterate over two or more sequences at the same time. This is useful when you need to perform an operation on each combination of elements from the different sequences.

- the outer for-loop iterates over the numbers list, and the inner for-loop iterates over the letters list. For each iteration of the outer loop, the inner loop iterates over the entire letters list, printing each combination of a number and a letter.

- we can nest for-loops to iterate over any number of sequences. Here's an example of a nested for-loop that iterates over a list of colors and a list of sizes, and prints each combination of a color and a size

In [14]:
numbers = [1, 2, 3]
letters = ['a', 'b', 'c']
for num in numbers:
    for letter in letters:
        print(num, letter)

1 a
1 b
1 c
2 a
2 b
2 c
3 a
3 b
3 c


In [16]:
colors = ['red', 'green', 'blue']
sizes = ['small', 'medium', 'large']
for color in colors:
    for size in sizes:
        print(color, size)

red small
red medium
red large
green small
green medium
green large
blue small
blue medium
blue large


In [17]:
list1 = [1, 2, 3]
list2 = [4, 5, 6]
pairs = []
for num1 in list1:
    for num2 in list2:
        pairs.append((num1, num2))
print(pairs)

[(1, 4), (1, 5), (1, 6), (2, 4), (2, 5), (2, 6), (3, 4), (3, 5), (3, 6)]


In [27]:
for i in range(1, 11):       # Printing multiplication table from 1 to 10:
    for j in range(1, 11):
        print(i * j, end="\t")
    print()

1	2	3	4	5	6	7	8	9	10	
2	4	6	8	10	12	14	16	18	20	
3	6	9	12	15	18	21	24	27	30	
4	8	12	16	20	24	28	32	36	40	
5	10	15	20	25	30	35	40	45	50	
6	12	18	24	30	36	42	48	54	60	
7	14	21	28	35	42	49	56	63	70	
8	16	24	32	40	48	56	64	72	80	
9	18	27	36	45	54	63	72	81	90	
10	20	30	40	50	60	70	80	90	100	


__3.While Loops__

- A while loop in programming is a control flow statement that allows you to execute a block of code repeatedly as long as a condition is true. The loop will continue to execute as long as the condition remains true.

- The condition is typically a Boolean expression that evaluates to true or false. The code inside the while loop is executed repeatedly as long as the condition is true. If the condition is false when the loop is first encountered, the code inside the loop will not be executed at all.

In [1]:
num = 1        #Print numbers from 1 to 5 using a while loop
while num <= 5:
    print(num)
    num += 1

1
2
3
4
5


In [2]:
num = 1       #Calculate the sum of all even numbers from 1 to 10 using a while loop
sum = 0

while num <= 10:
    if num % 2 == 0:
        sum += num
    num += 1

print(f"Sum of even numbers from 1 to 10: {sum}")

Sum of even numbers from 1 to 10: 30


In [5]:
password = "Manohar@1"   #Asking the user for a password and keep asking until the correct password is entered using a while loop
input_password = input("Enter password: ")

while input_password != password:
    print("Incorrect password!")
    input_password = input("Enter password: ")

print("Access granted!")

Enter password: Manohar@1
Access granted!


In [6]:
string = "Manohar"   #Reversing string using a while loop
reversed_string = ""
index = len(string) - 1

while index >= 0:
    reversed_string += string[index]
    index -= 1

print("Original string:", string)
print("Reversed string:", reversed_string)

Original string: Manohar
Reversed string: rahonaM


In [10]:
num = 5        #Calculating the factorial of a given number using a while loop
factorial = 1
i = 1

while i <= num:
    factorial *= i
    i += 1

print(f"Factorial of {num} is {factorial}")

Factorial of 5 is 120


In [11]:
#Print a multiplication table for a given number using a while loop
num = 7
i = 1

while i <= 10:
    print(num, "x", i, "=", num * i)
    i += 1

7 x 1 = 7
7 x 2 = 14
7 x 3 = 21
7 x 4 = 28
7 x 5 = 35
7 x 6 = 42
7 x 7 = 49
7 x 8 = 56
7 x 9 = 63
7 x 10 = 70


In [12]:
#Asking the user to guess a number between 1 and 10 and keep asking until the correct number is guessed using a while loop
number = 7
guess = int(input("Guess a number between 1 and 10: "))
while guess != number:
    print("Wrong guess, try again!")
    guess = int(input("Guess a number between 1 and 10: "))
print("Congratulations, you guessed the number!")

Guess a number between 1 and 10: 7
Congratulations, you guessed the number!


In [14]:
#Finding the largest number in a given list using a while loop
list1 = [10, 20, 4, 45, 99]
index = 0
largest = list1[index]

while index < len(list1):
    if list1[index] > largest:
        largest = list1[index]
    index += 1

print(f"The largest number in the list is: {largest}")

The largest number in the list is: 99


__Loop Controls:__

Loop controls are used to change the flow of execution within a loop in a program.loop controls are:
 - Break statement
 - Continue statement
 - Pass statement

In [1]:
#Break
for i in range(1, 11):
    if i == 5:
        break
    print(i)

1
2
3
4


In [2]:
#Continue statement
for i in range(1, 11):
    if i == 5:
        continue
    print(i)

1
2
3
4
6
7
8
9
10


In [3]:
#Pass statement
for i in range(1, 11):
    if i % 2 == 0:
        pass
    else:
        print(i)

1
3
5
7
9


In [4]:
#Nested loop with break: This example shows how to use a break statement to immediately terminate a nested loop:
for i in range(1, 4):
    for j in range(1, 4):
        if i == 2 and j == 2:
            break
        print(i, j)

1 1
1 2
1 3
2 1
3 1
3 2
3 3


In [5]:
#While loop with continue: This example shows how to use a continue statement to skip over iterations in a while loop:
i = 0
while i < 5:
    i += 1
    if i == 3:
        continue
    print(i)

1
2
4
5


In [6]:
#For loop with pass: This example shows how to use a pass statement to do nothing in a for loop:
fruits = ['apple', 'banana', 'orange', 'kiwi']
for fruit in fruits:
    if fruit == 'orange':
        pass
    else:
        print(fruit)

apple
banana
kiwi


__Comprehension:__

Comprehensions provide us with a short and concise way to construct new sequences (such as lists, set, dictionary etc.) using sequences which have been already defined.

- Comprehensions in Python are a concise way to create collections such as lists, sets, and dictionaries.

In [7]:
#List comprehension: This creates a new list based on an existing list or other iterable.
# Creating a list of squares of numbers from 1 to 10
squares = [x*x for x in range(1, 11)]
print(squares)

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


In [12]:
# Traditional method
even_list = []
for i in range(1,21):
    if i%2 == 0:
        even_list.append(i)

In [13]:
print(even_list)

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20]


In [14]:
# Comprehension method
even_list_compro = [i for i in range(1,21)]
print(even_list_compro)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]


In [8]:
#Set comprehension: This creates a new set based on an existing list or other iterable.
# Creating a set of even numbers from 1 to 10
evens = {x for x in range(1, 11) if x % 2 == 0}
print(evens)

{2, 4, 6, 8, 10}


In [9]:
#Dictionary comprehension: This creates a new dictionary based on an existing dictionary or other iterable.
# Creating a dictionary of squares of numbers from 1 to 10
squares = {x: x*x for x in range(1, 11)}
print(squares)

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100}


__Functions:__
- Functions in Python are blocks of code that can be called and executed repeatedly throughout a program. They are defined using the def keyword, followed by the function name, and any arguments that the function takes.
- functions are a powerful tool for organizing and reusing code in Python. They can take arguments, return values, and have optional arguments with default values. Functions can also be defined within other functions and can be passed as arguments to other functions, making them a versatile tool for solving complex programming problems.

In [16]:
def add_numbers(x, y):
    sum = x + y
    return sum

In this example, the def keyword is used to define the function add_numbers, which takes two arguments x and y. The body of the function calculates their sum and returns it using the return keyword.

In [17]:
#To call the function, you simply pass in values for the arguments, like this:
result = add_numbers(2, 3)
print(result)

5


In this example, we call the add_numbers function with arguments 2 and 3, and assign the result to the variable result. We then print the value of result, which is 5.

In [18]:
def greet(name, greeting='Hello'):
    print(greeting + ', ' + name + '!')

In [22]:
greet('Manohar')
greet('have a nice day', 'Good morning')

Hello, Manohar!
Good morning, have a nice day!


In this example, the greet function takes two arguments: name and greeting, with a default value of 'Hello'. If no value is specified for greeting when the function is called, it will use the default value.

In [23]:
def multiply_numbers(x, y):
    product = x * y
    return product

result = multiply_numbers(2, 3)
print(result)

6


In [24]:
# CALCULATOR FUNCTION
def calculator(a, b, operator):
    if operator == "+":
        print(f"The addition value is: {a+b}")
    elif operator == "-":
        print(f"The substraction value is: {a-b}")
    elif operator == "*":
        print(f"The multiplication value is: {a*b}")
    elif operator == "/":
        print(f"The division value is: {a/b}")
    else:
        print("error")

In [25]:
calculator(25, 45, "+")

The addition value is: 70


In [26]:
calculator(55, 20, "-")

The substraction value is: 75


In [27]:
calculator(5, 5, "*")

The multiplication value is: 10


In [28]:
calculator(55, 2, "/")

The division value is: 57


__Local and Global Variable:__


- __Local variables:__ A variable that is defined within a function is called a local variable. It can only be accessed within the function it was defined in. If you try to access it outside the function, you will get an error.

- __Global variables:__ A variable that is defined outside a function is called a global variable. It can be accessed from any part of the code, including all functions.

In [29]:
#Local variable
def my_function():
    x = 3
    print(x)

my_function()

3


In this example, the variable x is defined inside the function my_function, so it is a local variable. The variable x is only accessible within the function, so when the function is called, the value of x is printed to the console.

In [30]:
#Global variable 
x = 3

def my_function():
    print(x)

my_function()

3


n this example, the variable x is defined outside the function my_function, so it is a global variable. The variable x is accessible anywhere in the code, so when the function is called, the value of x is printed to the console.

In [31]:
#Local and global variable
x = 3

def my_function():
    x = 4
    print(x)

my_function()
print(x)

4
3


In this example, the variable x is defined both globally and locally. When the function my_function is called, the local variable x with value 4 is used inside the function, so that value is printed to the console. However, outside the function, the global variable x with value 3 is still accessible, so that value is printed to the console after the function call.

In [32]:
#Global variable modification
x = 3

def my_function():
    global x
    x = 4

my_function()
print(x)

4


In this example, the global keyword is used inside the function my_function to indicate that the global variable x should be modified. When the function is called, the value of the global variable x is changed to 4, so that value is printed to the console after the function call.

__Nested Functions:__

a nested function is a function defined inside another function. It is also called a local function because it is only accessible within the scope of the outer function.

In [33]:
def outer_function(x):
    def inner_function(y):
        return x + y
    return inner_function

closure = outer_function(5)
result = closure(3)
print(result)

8


In this example, outer_function returns inner_function which is defined inside it. The returned function inner_function has access to the variable x from the outer function, even after outer_function has finished executing. This is an example of a closure, where a function retains access to the variables in its enclosing scope.

In [34]:
def outer_function():
    x = 10
    def inner_function():
        nonlocal x
        x += 5
        print("The value of x inside inner function is:", x)
    inner_function()
    print("The value of x outside inner function is:", x)

outer_function()

The value of x inside inner function is: 15
The value of x outside inner function is: 15


In this example, inner_function is nested inside outer_function, and has access to the variable x defined in outer_function. By using the nonlocal keyword, we can modify the value of x from within inner_function. The output shows that the value of x is modified inside inner_function, and the modified value is also reflected outside of it.

In [35]:
def outer_function(x):
    def inner_function(y):
        return x + y
    return inner_function

closure1 = outer_function(5)
closure2 = outer_function(10)
result1 = closure1(3)
result2 = closure2(3)
print(result1, result2)

8 13


In this example, we define outer_function which returns inner_function. We call outer_function twice with different values for x, creating two closures closure1 and closure2. Each closure has its own copy of x from the respective call to outer_function. We then call each closure with the same argument 3, resulting in different return values based on the value of x for each closure.

__Lambda Function:__
    
Lambda functions are a type of anonymous function in Python. They are created using the keyword lambda and allow you to define a simple function without needing to use the def keyword.

In [36]:
double = lambda x: x * 2
result = double(5)
print(result) # Output: 10

10


In this example, we define a lambda function double that takes one argument x and returns x * 2. We then call the lambda function with the argument 5, which results in the output 10.

In [37]:
points = [(1, 2), (3, 1), (5, 10), (2, 7)]
sorted_points = sorted(points, key=lambda x: x[1])
print(sorted_points) # Output: [(3, 1), (1, 2), (2, 7), (5, 10)]

[(3, 1), (1, 2), (2, 7), (5, 10)]


In this example, we have a list of points, where each point is represented as a tuple of two integers. We want to sort the points based on their second coordinate (i.e., the y value), so we use the sorted function and pass a lambda function as the key argument. The lambda function takes one argument x, which is a point tuple, and returns the second element of the tuple (i.e., the y value). This results in the list of points being sorted by their y value.

In [39]:
def operate_on_numbers(numbers, operation):
    return [operation(x) for x in numbers]

numbers = [1, 2, 3, 4, 5]
squared_numbers = operate_on_numbers(numbers, lambda x: x ** 2)
cubed_numbers = operate_on_numbers(numbers, lambda x: x ** 3)
print(squared_numbers)
print(cubed_numbers)

[1, 4, 9, 16, 25]
[1, 8, 27, 64, 125]


In this example, we define a function operate_on_numbers that takes a list of numbers and a function operation as arguments. The function applies the operation function to each number in the list using a list comprehension, and returns the resulting list. We then call operate_on_numbers twice with different lambda functions as the operation argument, resulting in two lists of squared and cubed numbers, respectively.

__Functions as Arguments to Functions:__
    
- functions can be passed as arguments to other functions. This is a powerful feature that allows us to write higher-order functions, which are functions that operate on other functions.

In [40]:
def apply_operation(x, operation):
    return operation(x)

def square(x):
    return x ** 2

result = apply_operation(5, square)
print(result)

25


In this example, we define a function apply_operation that takes two arguments x and operation. The function applies the operation function to the input x and returns the result. We also define a function square that squares its input. We then call apply_operation with the arguments 5 and square, which results in the output 25 (i.e., square(5)).

In [41]:
def sum_numbers(numbers):
    return sum(numbers)

def apply_to_each(numbers, operation):
    return [operation(x) for x in numbers]

numbers = [1, 2, 3, 4, 5]
sum_of_squares = sum(apply_to_each(numbers, lambda x: x ** 2))
sum_of_cubes = sum(apply_to_each(numbers, lambda x: x ** 3))
print(sum_of_squares)
print(sum_of_cubes)

55
225


In this example, we define a function sum_numbers that takes a list of numbers and returns the sum of the numbers. We also define a function apply_to_each that takes a list of numbers and a function operation, and returns a new list with the operation applied to each element of the input list. We then call apply_to_each twice with different lambda functions as the operation argument, resulting in two lists of squared and cubed numbers, respectively. We use the sum function to compute the sum of each list, resulting in the output 55 and 225.

In [43]:
def apply_twice(x, operation):
    return operation(operation(x))

def square(x):
    return x ** 2

result = apply_twice(5, square)
print(result)

625


In this example, we define a function apply_twice that takes two arguments x and operation. The function applies the operation function to the input x twice and returns the result. We also define a function square that squares its input. We then call apply_twice with the arguments 5 and square, which results in the output 625 (i.e., square(square(5))).

__Object Oriented Programming__

- Object-oriented programming (OOP) is a programming paradigm that is based on the concept of objects. An object is an instance of a class, which is a template for creating objects. OOP is widely used in modern software development because it provides a way to organize and structure code in a modular and reusable way.

- In OOP, classes are used to encapsulate data and behavior (i.e., methods) into a single entity. Objects are created from classes and can interact with each other by sending messages. The three main pillars of OOP are encapsulation, inheritance, and polymorphism.

- __Class:__ A blueprint for creating objects, which defines a set of attributes and methods that the objects will have.

- __Object:__ An instance of a class, which has its own set of attributes and can perform methods defined in the class.

- __Encapsulation:__ The practice of hiding internal implementation details of an object from the outside world, and only exposing a well-defined interface for interacting with the object.

- __Inheritance:__ A mechanism for creating a new class that is a modified version of an existing class. The new class inherits attributes and methods from the parent class and can also have its own additional attributes and methods.

- __Polymorphism:__ The ability of different objects to respond to the same message (i.e., method call) in different ways. Polymorphism allows for more flexible and generic code, as it enables code to work with different objects that may have different implementations of the same method.

__1.Class and objects:__

- Classes and objects are fundamental concepts in object-oriented programming (OOP). OOP is a programming paradigm that emphasizes the use of objects that have properties (attributes) and behaviors (methods). In Python, a class is a blueprint for creating objects, and an object is an instance of a class.

- A class is defined using the class keyword followed by the name of the class. The body of the class contains the attributes and methods of the class. Attributes are variables that hold data, while methods are functions that perform actions.

- When an object is created from a class, it is called an instance of that class. The process of creating an object from a class is called instantiation. The __init__ method is a special method that is called when an object is instantiated. It is used to set the initial values of the object's attributes.

In [9]:
#Creating a Car class
class Car:
    def __init__(self, make, model, year):
        self.make = make
        self.model = model
        self.year = year
    
    def drive(self):
        print(f"The {self.year} {self.make} {self.model} is driving.")

In [10]:
car = Car("Toyota", "Camry", 2020)

In [11]:
car.drive()

The 2020 Toyota Camry is driving.


In this example, we define a Car class that has three attributes (make, model, and year) and one method (drive). The __init__ method is called when a Car object is created and sets the initial values of the attributes. The drive method is a simple method that prints a message to the console using the attributes of the object.

In [12]:
class laptop:
    def __init__(self, brand, model, year):
        self.brand = brand
        self.model = model
        self.year = year
    
    def working(self):
        print(f"The {self.year} {self.brand} {self.model} is working good.")

In [13]:
laptop = laptop("acer", "aspir 5", 2023)

In [14]:
laptop.working()

The 2023 acer aspir 5 is working good.


In this example, we define a laptop class that has three attributes (brand, model, and year) and one method (working). The __init__ method is called when a laptop object is created and sets the initial values of the attributes. The working method is a simple method that prints a message to the console using the attributes of the object.

In [15]:
#Creating a Person class
class Person:
    def __init__(self, name, age):
        self.name = name
        self.age = age
    
    def say_hello(self):
        print(f"Hello, my name is {self.name} and I am {self.age} years old.")

In [16]:
person = Person("Manohar", 22)

In [17]:
person.say_hello()

Hello, my name is Manohar and I am 22 years old.


In this example, we define a person class that has two attributes (name and age) and one method (say_hello). The __init__ method is called when a Person object is created and sets the initial values of the attributes. The say_hello method is a simple method that prints a message to the console using the attributes of the object.

__Inheritance, Encapsulation and Polymorphism__

In object-oriented programming, inheritance, encapsulation, and polymorphism are three fundamental concepts that help in creating more maintainable, modular, and flexible code.

__1.Inheritance:__

Inheritance is a mechanism that allows a new class to be based on an existing class. The new class, called the subclass or derived class, inherits the properties and behavior of the existing class, called the superclass or base class, and can also add its own properties and behavior. Inheritance helps in reducing code duplication and enables code reuse.

In [18]:
class Animal:
    def __init__(self, name, species):
        self.name = name
        self.species = species
    
    def make_sound(self):
        pass

class Dog(Animal):
    def __init__(self, name, breed):
        super().__init__(name, species="Dog")
        self.breed = breed
    
    def make_sound(self):
        print("Woof!")

In [20]:
my_dog = Dog("Buddy", "Golden Retriever")
print(my_dog.name) 
print(my_dog.species) 
print(my_dog.breed)
my_dog.make_sound()

Buddy
Dog
Golden Retriever
Woof!


In this example, we define an Animal class with an __init__ method and a make_sound method. We then define a Dog class that inherits from the Animal class and overrides the __init__ and make_sound methods. The super() function is used to call the __init__ method of the superclass.

__2.Encapsulation:__

Encapsulation is the practice of hiding the internal details of an object from the outside world and only exposing a public interface that can be used to interact with the object. This helps in keeping the code modular, reducing coupling, and improving maintainability. In Python, encapsulation is achieved through the use of private and protected attributes and methods. Private attributes and methods are denoted by a double underscore prefix, while protected attributes and methods are denoted by a single underscore prefix.

In [22]:
class BankAccount:
    def __init__(self, account_number, balance):
        self.__account_number = account_number
        self.__balance = balance
    
    def deposit(self, amount):
        self.__balance += amount
    
    def withdraw(self, amount):
        if amount <= self.__balance:
            self.__balance -= amount
        else:
            print("Insufficient funds.")
    
    def get_balance(self):
        return self.__balance

In [24]:
my_account = BankAccount("123456789", 1000)
print(my_account.get_balance())
my_account.deposit(500)
print(my_account.get_balance())
my_account.withdraw(200)
print(my_account.get_balance())
my_account.withdraw(1500)

1000
1500
1300
Insufficient funds.


In this example, we define a BankAccount class with an __init__ method, deposit, withdraw, and get_balance methods. The account_number and balance attributes are encapsulated using double underscores to make them private. The public interface for interacting with the object is through the deposit, withdraw, and get_balance methods.

__3.Polymorphism:__

Polymorphism is the ability of objects of different classes to be used interchangeably. This is achieved through the use of inheritance and method overriding. Polymorphism helps in creating more flexible and reusable code.

In [25]:
class Animal:
    def make_sound(self):
        pass

class Dog(Animal):
    def make_sound(self):
        print("Woof!")

class Cat(Animal):
    def make_sound(self):
        print("Meow!")

def make_animal_sound(animal):
    animal.make_sound()

dog = Dog()
cat = Cat()

make_animal_sound(dog)
make_animal_sound(cat)

Woof!
Meow!


In this example, we define an Animal class with a make_sound method. We then define Dog and Cat classes that inherit from the Animal class and override the make_sound method. We also define a `make

__Error handling__

- Error handling in Python refers to the process of catching and handling errors (also known as exceptions) that can occur during the execution of a program. Exceptions are raised by the Python interpreter when it encounters an error that it can't handle, such as trying to divide by zero, accessing an undefined variable, or calling a method on an object that doesn't support it.

- Python provides several built-in mechanisms for error handling, including the try-except statement, which allows you to catch and handle specific types of exceptions. You can also use a generic except statement to catch any type of exception, or a finally statement to execute code that should always run, regardless of whether an exception was raised or not.

- In addition to these built-in mechanisms, Python also provides ways to raise your own exceptions using the raise statement, and to create custom exception classes that can be used to handle specific types of errors in your program.

- In Python, there are two main types of errors: syntax errors and exceptions.

__1.Syntax errors:__

These errors occur when the Python interpreter is unable to parse your code because it violates the rules of the Python language. Syntax errors are usually detected when you try to run your code and cause the interpreter to display an error message. Examples of syntax errors include misspelled keywords or variable names, missing colons or parentheses, and incorrect indentation.

__2.Exceptions:__

These errors occur when your code is syntactically correct, but encounters a problem during execution. Exceptions can be caused by a wide variety of problems, such as dividing by zero, trying to access an undefined variable, or calling a method on an object that doesn't support it. When an exception is raised, the Python interpreter stops executing your program and displays an error message. Exceptions can be caught and handled using the try-except statement.

__Try and Except__

In [26]:
# Handling a division by zero error
num1 = 10
num2 = 0

try:
    result = num1 / num2
except ZeroDivisionError:
    print("Cannot divide by zero!")

Cannot divide by zero!


In [27]:
#Handling a file not found error
filename = "nonexistentfile.txt"

try:
    with open(filename, "r") as f:
        data = f.read()
except FileNotFoundError:
    print(f"Error: {filename} not found!")

Error: nonexistentfile.txt not found!


In [28]:
#Handling a value error
user_input = "abc"

try:
    num = int(user_input)
except ValueError:
    print(f"Error: {user_input} is not a valid integer!")

Error: abc is not a valid integer!


__File Handling__

File handling in Python refers to the process of creating, reading, updating, and deleting files on a computer's file system. Python provides a built-in file handling module called os and io for this purpose. 

In [31]:
#Opening and reading a file
# open a file in read mode
file = open("manohar.txt", "r")

# read the contents of the file
data = file.read()

# print the data
print(data)

# close the file
file.close()

hi
good
morning
have
a
nice
day


In [39]:
#Writing to a file
# open a file in write mode
file = open("manu.txt", "w")

# write to the file
file.write("happy coding!")

# close the file
file.close()

In [40]:
file = open("manu.txt", "r")

# read the contents of the file
data = file.read()

# print the data
print(data)

# close the file
file.close()

happy coding!


In [41]:
#Appending to a file
# open a file in append mode
file = open("manu.txt", "a")

# append to the file
file.write("happy coding!")

# close the file
file.close()

In [42]:
file = open("manu.txt", "r")

# read the contents of the file
data = file.read()

# print the data
print(data)

# close the file
file.close()

happy coding!happy coding!


In [46]:
# check if a file exists
import os
if os.path.exists("manu.txt"):
    print("The file exists.")
else:
    print("The file does not exist.")

The file exists.


In [47]:
# delete a file
import os
os.remove("example.txt")

In [48]:
import os
if os.path.exists("example.txt"):
    print("The file exists.")
else:
    print("The file does not exist.")

The file does not exist.
